runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp

```diff
@@ -653,10 +653,156 @@ TR::Register *TR::ARM64PrivateLinkage::buildDirectDispatch(TR::Node *callNode)
+   TR::Register *x9 = dependencies->searchPreConditionRegister(TR::RealRegister::x9);
+
+   TR::SymbolReference *methodSymRef = callNode->getSymbolReference();
+   TR::MethodSymbol *methodSymbol = methodSymRef->getSymbol()->castToMethodSymbol();
+   TR::LabelSymbol *doneLabel = NULL;
+
+   TR::Instruction *gcPoint;
+
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(comp()->fe());
+
+   // Computed calls
+   //
+   if (methodSymbol->isComputed())
+      {
+      // ToDo: Implement this path (Eclipse OpenJ9 Issue #7023)
+      comp()->failCompilation<TR::AssertionFailure>("ComputedCall");
+      }
+
+   // Virtual and interface calls
+   //
+   TR_ASSERT(methodSymbol->isVirtual() || methodSymbol->isInterface(), "Unexpected method type");
+
+   void *thunk = fej9->getJ2IThunk(methodSymbol->getMethod(), comp());
+   if (!thunk)
+      thunk = fej9->setJ2IThunk(methodSymbol->getMethod(), TR::ARM64CallSnippet::generateVIThunk(callNode, argSize, codeGen), comp());
+
+   if (methodSymbol->isVirtual())
+      {
+      TR::MemoryReference *tempMR;
+      if (methodSymRef->isUnresolved() || comp()->compileRelocatableCode())
+         {
+         TR::Register *vftReg = evaluateUpToVftChild(callNode, codeGen);
+         TR::addDependency(dependencies, vftReg, TR::RealRegister::NoReg, TR_GPR, codeGen);
+
+         doneLabel = generateLabelSymbol(codeGen);
+
+         TR::LabelSymbol *vcSnippetLabel = generateLabelSymbol(codeGen);
+         TR::ARM64VirtualUnresolvedSnippet *vcSnippet =
+            new (trHeapMemory())
+            TR::ARM64VirtualUnresolvedSnippet(codeGen, callNode, vcSnippetLabel, argSize, doneLabel, (uint8_t *)thunk);
+         codeGen->addSnippet(vcSnippet);
+
+         // The following instructions are modified by _virtualUnresolvedHelper
+         // in aarch64/runtime/PicBuilder.spp to load the method pointer in x9
+         generateTrg1ImmInstruction(codeGen, TR::InstOpCode::movzx, callNode, x9, 0);
+         generateTrg1ImmInstruction(codeGen, TR::InstOpCode::movkx, callNode, x9, TR::MOV_LSL16);
+         generateTrg1ImmInstruction(codeGen, TR::InstOpCode::movkx, callNode, x9, TR::MOV_LSL32);
+         generateTrg1ImmInstruction(codeGen, TR::InstOpCode::movkx, callNode, x9, TR::MOV_LSL48);
+         tempMR = new (trHeapMemory()) TR::MemoryReference(vftReg, 0, codeGen);
+         generateTrg1MemInstruction(codeGen, TR::InstOpCode::ldrimmx, callNode, x9, tempMR);
+         gcPoint = generateLabelInstruction(codeGen, TR::InstOpCode::b, callNode, vcSnippetLabel, dependencies);
+         }
+      else
+         {
+         int32_t offset = methodSymRef->getOffset();
+         TR::Register *vftReg = evaluateUpToVftChild(callNode, codeGen);
+         TR::addDependency(dependencies, vftReg, TR::RealRegister::NoReg, TR_GPR, codeGen);
+
+         // jitVTableIndex() in oti/JITInterface.hpp assumes the instruction sequence below
+         loadConstant64(codeGen, callNode, (int64_t)offset, x9);
```

Comment:
Won't `loadConstant64` always generate 4 instructions if the offset isn't 0?  But the code in `jitVTableIndex` expects only one instruction.

Even if the code in `jitVTableIndex` will just work because it happens to pick up the LSB instruction of the constant construction, I don't think we should be emitting 4 instructions here.

Reply:
`loadConstant64()` generates only one `movnx` instruction when the offset is a negative value within 16-bit range.
I admit I haven't tested the case that requires two or more instructions to load the index into `x9`.

Reply:
At the very least, on the resolved path where you know the vtable index, can you add an assert to be sure it is within range?  In the near future, I think we need to look into a different scheme for loading constants in some contexts (e.g. using the PC relative loads) to make this code more solid.

Reply:
Stopped using `loadConstant64()` here, and instead explicitly generate a single `movnx` instruction or longer instruction sequence (if wider than 16 bit).  Also changed JITInterface.hpp to reflect it.
The "wider index" path is tested locally by disabling the single `movnx` path intentionally.